sensors: move differential pressure aggregation to sensors/airspeed WorkItem

dagar opened the pull request:
This gets differential_pressure/airspeed handling out of the main `sensors` loop and into a standalone WorkItem that's only allocated if you have a differential pressure sensor. It also gets more of the pieces in place for properly handling multiple sensors. Most of the structure mirrors what's been done for barometers and magnetometers.

New parameters.
 - SENS_DPRES_RATE to control publication rate
     - data is averaged and only published at SENS_DPRES_RATE, with the timestamp updated appropriately to reflect the age of the data
     - default 10 Hz
 - SENS_DPRES_MODE to control publishing only 1 or N `airspeed` msgs

dagar committed:
sensors: move differential pressure aggregation to sensors/airspeed WorkItem

dagar committed:
ekf2: use airspeed timestamp_sample

dagar committed:
mavlink: HIL publish airspeed with timestamp_sample

sfuhrer reviewed (commented):
At a first glance looks like a big step forward, and is certainly cleaner and better structured.
I'm wondering though if having doing duplicated airspeed validation and selection of the best in here and then inside the airspeed_selector again is the best option. Maybe we should do all the voting inside the airspeed_selector, or find a way to propagate the vote to the airspeed_selector, and there do the selection. That would mean we could always publish all airspeeds.

dagar commented:
> Maybe we should do all the voting inside the airspeed_selector, or find a way to propagate the vote to the airspeed_selector, and there do the selection. That would mean we could always publish all airspeeds.

I think that could make sense, with a couple questions/caveats.
1. What do we send to ekf2?
   - In the future would we ever want to send different airspeeds to different ekf2 instances (https://github.com/PX4/Firmware/pull/14650), or does the airspeed selector completely negate the need?
2. The direction of the `sensors` module is to be the hub for all raw data regardless of the source (I2C/SPI/UART/UAVCAN/Mavlink) where  everything is processed (calibrations applied, user facing config, etc) before being passed on to the rest of the system either directly or after "voting" (simple health check and priority, potentially median select). For differential pressure that's going to expand slightly with calibrations per sensor (with device ids), and either a param for a user to enable/disable per sensor or perhaps a configurable priority like what what we have for accel/gyro/mag now. https://github.com/PX4/Firmware/issues/15683

dagar commented on src/modules/airspeed_selector/airspeed_selector_main.cpp:
```diff
@@ -303,9 +306,9 @@ AirspeedModule::Run()
-		input_data.timestamp = _time_now_usec;
+		// Prepare data for airspeed_validator
+		airspeed_validator_update_data input_data{};
+		input_data.timestamp = _vehicle_local_position.timestamp_sample;
```
It's minor, but it minimizes most of the jitter in your delta time (the only use I saw at a glance) using the time that corresponds with the input data rather than when the module happened to actually run.

dagar commented:
https://github.com/PX4/Firmware/issues/12237#issuecomment-702779028

sfuhrer commented:
>     1. What do we send to ekf2?
>        
>        * In the future would we ever want to send different airspeeds to different ekf2 instances (#14650), or does the airspeed selector completely negate the need?
I think that's a key point to decide first, as a lot of things depend on it. What is currently planned for GPS and baro? Do you also envision that there will be different EKF instances for a dual GPS setup? Doing it for all the sensors of "secondary" importance would probably lead to too many instances, so should we clearly categorize the sensors into two groups (accel, gyro, mag) and (GPS, baro, airspeed, distance, ..) and then only allow for multiple EKFs for group 1, while we do validation and selection for group 2 with other means (e.g. like we do it now for airspeed)?
If we don't have separate instances of EKF per airspeed sensor the dependencies will get nasty though - EKF requires knowledge about airpspeed health to decide when to fuse it and when not, and the airspeed validator in turn would require attitude and local position from EKF. At least if we want to keep the airspeed innovation check.

dagar commented:
> What is currently planned for GPS and baro? Do you also envision that there will be different EKF instances for a dual GPS setup? Doing it for all the sensors of "secondary" importance would probably lead to too many instances, so should we clearly categorize the sensors into two groups (accel, gyro, mag) and (GPS, baro, airspeed, distance, ..) and then only allow for multiple EKFs for group 1, while we do validation and selection for group 2 with other means (e.g. like we do it now for airspeed)?

That's a good question with no real answer at the moment other than I think we'll need to be flexible (and realistic) to handle a wide ranges of setups. For all of these options (with the possible exception of distance) they'll be configure at the `sensors` hub to either pick a primary (highest priority, blended, median select, etc) or pass on all instances. Another option I was thinking about was passing on a certain subset like the top 2 mags on a system that has 2 external and 1 internal. We could almost completely ignore the internal unless one of the externals is completely lost (died or disconnected in flight).

Newer boards with an STM32H7 might be able to afford the full matrix (if we raise ORB_MULTI_MAX_INSTANCES beyond 10), but on STM32F7 a more realistic limit is probably 6. On older boards (STM32F4) we could afford a couple, but memory is tight enough that it's not going to be on by default.

I'm also hoping in the future we'll start seeing actual triplex configurations for certain sensors so that we can simply median select at the sensors hub.

sfuhrer commented:
> > What is currently planned for GPS and baro? Do you also envision that there will be different EKF instances for a dual GPS setup? Doing it for all the sensors of "secondary" importance would probably lead to too many instances, so should we clearly categorize the sensors into two groups (accel, gyro, mag) and (GPS, baro, airspeed, distance, ..) and then only allow for multiple EKFs for group 1, while we do validation and selection for group 2 with other means (e.g. like we do it now for airspeed)?
> 
> That's a good question with no real answer at the moment other than I think we'll need to be flexible (and realistic) to handle a wide ranges of setups. For all of these options (with the possible exception of distance) they'll be configure at the `sensors` hub to either pick a primary (highest priority, blended, median select, etc) or pass on all instances. Another option I was thinking about was passing on a certain subset like the top 2 mags on a system that has 2 external and 1 internal. We could almost completely ignore the internal unless one of the externals is completely lost (died or disconnected in flight).
> 
> Newer boards with an STM32H7 might be able to afford the full matrix (if we raise ORB_MULTI_MAX_INSTANCES beyond 10), but on STM32F7 a more realistic limit is probably 6. On older boards (STM32F4) we could afford a couple, but memory is tight enough that it's not going to be on by default.
> 
> I'm also hoping in the future we'll start seeing actual triplex configurations for certain sensors so that we can simply median select at the sensors hub.

Okay, it then sounds like proceeding as you begun makes most sense, with some basic airspeed validity checks in sensors, and the publication of the best or all. Assuming that one has two airspeed sensors and a very reliable mag, would it in your perspective make sense run the 6 EKF instances for (2 accel, 2 gyro, 2 airspeeds)?

dagar commented:
> Assuming that one has two airspeed sensors and a very reliable mag, would it in your perspective make sense run the 6 EKF instances for (2 accel, 2 gyro, 2 airspeeds)?

A couple points.
 - We keep accel & gyro paired as a single IMU (which is physically true for the vast majority).
 - Things like mag are where I want to be realistic. If you have one or more garbage internal mags that are unusably bad let's not even use them.

I'm not sure what to think about the potential coupling between multiple raw airpseed, multi-ekf2, and the airspeed selector. I think it may be a worthwhile exercise to go through a simple fault tree analysis for a couple non-contrived scenarios (sensor combinations + specific board resource limitations). In particular I'm wondering about the best option overall in a loss of GPS scenario where we want to be able to dead-reckon (airspeed assisted) as long as possible to get home or emergency land.

sfuhrer commented:
> I think it may be a worthwhile exercise to go through a simple fault tree analysis for a couple non-contrived scenarios (sensor combinations + specific board resource limitations). In particular I'm wondering about the best option overall in a loss of GPS scenario where we want to be able to dead-reckon (airspeed assisted) as long as possible to get home or emergency land.

A fault tree analysis would ofc be very reasonable, mostly as we're still tapping pretty much in the dark concerning airspeed failure types/causes/results. 

Meanwhile I'll here just drop a couple of guesses/assumptions:
If we for now do not plan to have GPS and airspeed checked via multi-EKF, but agree that failures of these two external sensors are most likely (compared to baro mainly), then we could actually already run basic validity checks for them that are completely separated (and are also separated from the IMU/mag fallback logic that comes in with multiple EKFs):
- airspeed: check for data missing and airspeed < stall speed (If you have your cruise speed close to stall, which often is the case, I would assume that also a partially clogged airspeed could be detected that way)
- GPS: gives itself feedback about accuracy, number of sats connected etc.

With these checks, then we can do fallbacks to alternative airspeed/GPS, and if that's not available, then move on to airspeed-less cruise (in case of airspeed failure), or airspeed-assisted dead-reckoning (in case of GPS failure), or to a constant bank angle/emergency landing if we get bot GPS and airspeed failures at the same time.

Question: could we do checks for airspeed<stall inside the sensors hub? If yes, then the only reason we still need an airspeed_selector would be to have the check for airspeed to wind innovations, plus (potentially, as not yet implemented) compare the innovations of multiple sensors and pick the one with the smallest/most consistent innovations.

dagar commented:
> Question: could we do checks for airspeed<stall inside the sensors hub? If yes, then the only reason we still need an airspeed_selector would be to have the check for airspeed to wind innovations, plus (potentially, as not yet implemented) compare the innovations of multiple sensors and pick the one with the smallest/most consistent innovations.

Yes as long as the data flow is unidirectional (avoid coupling with estimators, etc). The EKF backend (ecl/EKF) has several checks that I'd like to pull back to sensors (many of the GPS checks, IMU vibration metrics, etc). 

In general I'm trying to structure sensor health as a hierarchy that rolls into overall reporting to the system (and operator). Starting from the individual sensor reporting transaction errors, to the sensors hub performing simple checks for timeouts, stuck data, comparing readings from multiple, then finally use in each estimator instance.

sfuhrer commented:
> In general I'm trying to structure sensor health as a hierarchy that rolls into overall reporting to the system (and operator). Starting from the individual sensor reporting transaction errors, to the sensors hub performing simple checks for timeouts, stuck data, comparing readings from multiple, then finally use in each estimator instance.

Sounds good. In that case it sounds to me like we should think about removing the airspeed_selector completely once we have the sensors hub in that shape.

dagar commented:
@sfuhrer good to go?

sfuhrer commented:
> @sfuhrer good to go?

Yes from my side. Let's drive the whole airpseed re-architecture from lower to higher level - meaning that this looks like a much cleaner implementation on the sensors level, and we can after it start discussions about it's usage in the EKF resp airspeed selector module. 
Did you test it with multiple sensors, inclusive failover?

dagar commented:
> Didn't find any obvious error. Did you test it with multiple sensors, inclusive failover?

No, I wasn't really considering multi-airspeed "supported" until the device ids, orb messages, parameters, etc are in place, but that's partially my perspective of seeing more users of the non-Sensirion options.

dagar committed:
differential_pressure/ets: set invalid temperature to NAN

dagar committed:
delete drv_airspeed.h (all IOCTLs), airspeed driver lib

 - differential pressure now processed downstream like other sensor data (accel/gyro/mag/etc)
 - rename msg differential_pressure -> sensor_differential_pressure
 - add device id and timestamp_sample to all messages
 - calibration performend directly on raw data and SENS_DPRES_OFF parameter updated
 - remove Airspeed base class from existing differential pressure drivers
 - name differential pressure drivers consistently (no _airspeed suffix)

dagar commented:
Next steps after this settles will be adding differential pressure calibration parameters per sensor (device id). This will also include a scale factor and optionally things like tube length.

sfuhrer commented on src/modules/airspeed_selector/airspeed_selector_main.cpp:
```diff
@@ -540,19 +543,29 @@ void AirspeedModule::select_airspeed_and_publish()
+
+		if (_valid_airspeed_index >= 0 && _valid_airspeed_index <= MAX_NUM_AIRSPEED_SENSORS) {
+
+			for (auto &sub : _airspeed_subs) {
```
I lost some context since the last review - what is the unregistration/registration of the airspeed_subs for here?

dagar replied:
This is so that `airspeed_selector` is scheduled to run immediately after the current primary airspeed publishes.

sfuhrer replied:
aha, good :+1:

dagar committed:
drivers/differential_pressure: split ms4515 and ms4525

dagar committed:
move analog differential pressure to standalone optional driver

MartinStokelj reviewed (changes requested):
I think you need to change collect() function, the sensor 4515 has different calculate values.

`
	int
	MEASAirspeed::collect4515()
	{
		int	ret = -EIO;

		/* read from the sensor */
		uint8_t val[4] = {0, 0, 0, 0};


		perf_begin(_sample_perf);

		ret = transfer(nullptr, 0, &val[0], 4);

		if (ret < 0) {
			perf_count(_comms_errors);
			perf_end(_sample_perf);
			return ret;
		}

		uint8_t status = (val[0] & 0xC0) >> 6;

		switch (status) {
		case 0:
			// Normal Operation. Good Data Packet
			break;

		case 1:
			// Reserved
			return -EAGAIN;

		case 2:
			// Stale Data. Data has been fetched since last measurement cycle.
			return -EAGAIN;

		case 3:
			// Fault Detected
			perf_count(_comms_errors);
			perf_end(_sample_perf);
			return -EAGAIN;
		}

		int16_t dp_raw = 0, dT_raw = 0;
		dp_raw = (val[0] << 8) + val[1];
		/* mask the used bits */
		dp_raw = 0x3FFF & dp_raw;
		dT_raw = (val[2] << 8) + val[3];
		dT_raw = (0xFFE0 & dT_raw) >> 5;

		// dT max is almost certainly an invalid reading
		if (dT_raw == 2047) {
			perf_count(_comms_errors);
			return -EAGAIN;
		}

		float temperature = ((200.0f * dT_raw) / 2047) - 50;

		// Calculate differential pressure. As its centered around 8000
		// and can go positive or negative
		const float P_min = -4.0f;
		const float P_max = 4.0f;
		const float H2O_to_Pa = 249.09f;
		/*
		  this equation is an inversion of the equation in the
		  pressure transfer function figure on page 4 of the datasheet

		  We negate the result so that positive differential pressures
		  are generated when the bottom port is used as the static
		  port on the pitot and top port is used as the dynamic port
		 */
		float diff_press_H2O = -((dp_raw - 0.05f * 16383) * (P_max - P_min) / (0.9f * 16383) + P_min);
		float diff_press_pa_raw = diff_press_H2O * H2O_to_Pa;

			/*
		  With the above calculation the MS4515 sensor will produce a
		  positive number when the top port is used as a dynamic port
		  and bottom port is used as the static port
		 */

		if (PX4_ISFINITE(diff_press_pa_raw)) {
			differential_pressure_s report{};

			report.error_count = perf_event_count(_comms_errors);
			report.temperature = temperature;
			report.differential_pressure_filtered_pa = _filter.apply(diff_press_pa_raw) - _diff_pres_offset;
			report.differential_pressure_raw_pa = diff_press_pa_raw - _diff_pres_offset;
			report.device_id = _device_id.devid;
			report.timestamp = hrt_absolute_time();

			_airspeed_pub.publish(report);
		}

		ret = OK;

		perf_end(_sample_perf);

		return ret;
	}
`
